### DISCUSSION TITLE
Troubleshooting Failed Bitcoin-related Test Cases in Compiling lnd from Source

### DISCUSSION THREAD

**Discussion by mohamedawnallah** [lightningnetwork/lnd#8272] [Q&A] [answered]
Hey everyone,

I'm in the process of compiling lnd from the source by following the instructions in the [INSTALL.md doc](https://github.com/lightningnetwork/lnd/blob/master/docs/INSTALL.md#installing-lnd-from-source). Everything seemed to install correctly, but I encountered issues when running the test cases using `make check`. All the tests passed except the ones associated with `bitcoind`. These failures seem to be related to the configuration and connection issues with `bitcoind`.

I've placed the necessary bitcoin-core binaries (such as `bitcoind`) in `/usr/local/bin` located in the `$PATH`. These binaries were downloaded from the [latest version of bitcoin core 26.0](https://bitcoincore.org/bin/bitcoin-core-26.0/), specifically using the version `bitcoin-26.0-arm64-apple-darwin.tar.gz`.

The error log I received after running `make check` includes messages like:
```
ok  	github.com/lightningnetwork/lnd/lnwallet/btcwallet	(cached)
--- FAIL: TestFilteredChainView (10.42s)
    interface_test.go:1010: Testing 'bitcoind_zmq' implementation of FilteredChainView
    interface_test.go:1022: unable to make chain view: unable to establish connection to bitcoind: method did not return within the timeout
FAIL
FAIL	github.com/lightningnetwork/lnd/routing/chainview	10.435s
FAIL
ok  	github.com/lightningnetwork/lnd/routing	(cached)
ok  	github.com/lightningnetwork/lnd/autopilot	(cached)
ok  	github.com/lightningnetwork/lnd/brontide	(cached)
ok  	github.com/lightningnetwork/lnd/cert	(cached)
--- FAIL: TestHistoricalConfDetailsTxIndex (0.00s)
    --- FAIL: TestHistoricalConfDetailsTxIndex/rpc_polling_enabled (10.32s)
        bitcoind_test.go:126: unable to establish connection to bitcoind: method did not return within the timeout
    --- FAIL: TestHistoricalConfDetailsTxIndex/rpc_polling_disabled (10.36s)
        bitcoind_test.go:126: unable to establish connection to bitcoind: method did not return within the timeout
--- FAIL: TestHistoricalConfDetailsNoTxIndex (0.00s)
    --- FAIL: TestHistoricalConfDetailsNoTxIndex/rpc_polling_enabled (10.37s)
        bitcoind_test.go:222: unable to establish connection to bitcoind: method did not return within the timeout
    --- FAIL: TestHistoricalConfDetailsNoTxIndex/rpc_polling_disabled (10.40s)
        bitcoind_test.go:222: unable to establish connection to bitcoind: method did not return within the timeout
FAIL
FAIL	github.com/lightningnetwork/lnd/chainntnfs/bitcoindnotify	20.776s
FAIL
```

My `bitcoin.conf` file is configured for `regtest` as follows:
```conf
# Generated by https://jlopp.github.io/bitcoin-core-config-generator/

# This config should be placed in following path:
# $HOME/Library/Application Support/Bitcoin/bitcoin.conf

regtest=1
server=1
daemon=1
debug=1
printtoconsole=1
debuglogfile=debug.log

# [Sections]
# Most options automatically apply to mainnet, testnet, and regtest networks.
# If you want to confine an option to just one network, you should add it in the relevant section.
# EXCEPTIONS: The options addnode, connect, port, bind, rpcport, rpcbind and wallet
# only apply to mainnet unless they appear in the appropriate section below.

# Options only for mainnet
[main]
prune=1000MB
rpcport=8332
rpcbind=0.0.0.0 # Note: For Testing purposes
rpcallowip=0.0.0.0/0 # Note: For Testing purposes
rpcuser=mainnet # Note: For Testing purposes
rpcpassword=mainnet # Note: For Testing purposes

# Options only for testnet
[test]

# Options only for regtest
[regtest]
fallbackfee=0.000001
port=18444
noconnect=1
dnsseed=0
dns=0
upnp=0
onlynet=ipv4
rpcport=18443
rpcbind=0.0.0.0 # Note: For Testing purposes
rpcallowip=0.0.0.0/0 # Note: For Testing purposes
rpcuser=regtest # Note: For Testing purposes
rpcpassword=regtest # Note: For Testing purposes
zmqpubrawblock=tcp://127.0.0.1:28332
zmqpubrawtx=tcp://127.0.0.1:28333

```

I'm running this on macOS m1 with the following system information:
```
Darwin Mohameds-MacBook-Air.local 23.0.0 Darwin Kernel Version 23.0.0: Fri Sep 15 14:41:34 PDT 2023; root:xnu-10002.1.13~1/RELEASE_ARM64_T8103 arm64
```

Any suggestions or insights on how to resolve these bitcoind-related test failures would be greatly appreciated! 🙏

**Comment by guggero** [comment]
Hmm, interesting. This might be an issue with `v26.0` of `bitcoind`. I just ran the same test locally with `v26.0` and got the same failure.
Or maybe our tests are just flaky (unfortunately some still are). Let me re-try with an earlier version of `bitcoind` to get some more info.

**Reply by mohamedawnallah** [reply]
Hey @guggero Any updates on this?

**Reply by guggero** [reply]
Yeah, it looks like on local runs those tests are just quite flakey. I got similar errors with versions from `v24.0` up to `v26.0`. But it looks like on GitHub with `v23.0` things pass but fail with `v26.0`: https://github.com/lightningnetwork/lnd/actions/runs/7186892040/job/19573287713?pr=8273
So I think this needs further investigation.

What's your intent with running the checks? Just out of interest or making sure a Pull Request passes? If the second is the case, I guess you can just create the PR and look at the results from GitHub CI.

**Reply by mohamedawnallah** [reply]
@guggero I’ve tried to downgrade to bitcoind `v23.0` as I checked it in the [main CI yaml file of lnd repo](https://github.com/lightningnetwork/lnd/blob/ac9ca024a4bfb2890e2cd450666bd94d221fe172/.github/workflows/main.yml#L24). It does give me another error regards `zmq`:
```
--- FAIL: TestFilteredChainView (26.66s)
    interface_test.go:1010: Testing 'bitcoind_zmq' implementation of FilteredChainView
    --- FAIL: TestFilteredChainView/bitcoind_zmq:_filter_block_disconnected (10.32s)
        interface_test.go:508:
            	Error Trace:	/Users/mohamedawnallah/Desktop/lnd/routing/chainview/interface_test.go:508
            	            				/Users/mohamedawnallah/Desktop/lnd/routing/chainview/interface_test.go:1031
            	Error:      	Received unexpected error:
            	            	unable to establish connection to bitcoind: unable to subscribe for zmq block events: dial unix //var/folders/nw/qczy4tjj333g09tcy197q3s40000gn/T/TestFilteredChainViewbitcoind_zmq_filter_block_disconnected1054260908/001/blocks.socket: connect: invalid argument
            	Test:       	TestFilteredChainView/bitcoind_zmq:_filter_block_disconnected
            	Messages:   	unable to create chain view
    interface_test.go:1010: Testing 'bitcoind_polling' implementation of FilteredChainView
    interface_test.go:1010: Testing 'p2p_neutrino' implementation of FilteredChainView
    interface_test.go:1010: Testing 'btcd_websockets' implementation of FilteredChainView
FAIL
FAIL	github.com/lightningnetwork/lnd/routing/chainview	26.684s
FAIL
```

Here are the steps I’ve done:
1) Downloaded `bitcoin-23.0-arm64-apple-darwin.tar.gz` from [bitcoin-core v23.0](https://bitcoincore.org/bin/bitcoin-core-23.0/bitcoin-23.0-arm64-apple-darwin.tar.gz) official download page.
2) Moved the binary files `bitcoind` in `/usr/local/bin` which included in the `$PATH`
3) Modified the `bitcoin.conf` configuration as you see in the main discussion and included it in the `regtest` mode options
```
[regtest]
...
zmqpubrawblock=tcp://127.0.0.1:28332
zmqpubrawtx=tcp://127.0.0.1:28333
```
4) Finally I run it using `make check`

Any suggestions or insights on resolving this issue would be greatly appreciated! 🙏

**Reply by mohamedawnallah** [reply]
> What's your intent with running the checks? Just out of interest or making sure a Pull Request passes? If the second is the case, I guess you can just create the PR and look at the results from GitHub CI.

Yeah I’ve looked with your Draft PR: [GitHub use bitcoind v26.0 for CI ](https://github.com/lightningnetwork/lnd/pull/8273)and it fails with bitcoind `v26.0`. Mainly I’d like to contribute to `lnd` repo so I started looking for `INSTALL.md` but If it is not necessary It is fine to go with bitcoind `v23.0`.

**Reply by mohamedawnallah** [reply]
I’ve found that my `bitcoind` already supports `ZeroMQ` by following this bitcoin stack exchange discussion [“How can I tell if bitcoind was compiled with ZeroMQ support"](https://bitcoin.stackexchange.com/a/102350) then I run `bitcoin-cli getzmqnotifications` command and it gives me this kind of response. That means I already have ZMQ supported, right ?:
```
➜  lnd git:(master) ✗ bitcoin-cli getzmqnotifications
[
  {
    "type": "pubrawblock",
    "address": "tcp://127.0.0.1:28332",
    "hwm": 1000
  },
  {
    "type": "pubrawtx",
    "address": "tcp://127.0.0.1:28333",
    "hwm": 1000
  }
]
```
After that still, I’m getting the same error as mentioned above after running `make check` with bitcoind `v23.0`. I don’t know if anything I’m doing wrong or setup incorrectly. Thanks!

**Reply by guggero** [reply]
I haven't seen this `dial unix //var/.../blocks.socket: connect: invalid argument` before. But after googling a bit, I found this: https://github.com/containers/podman/issues/13609#issuecomment-1076366318

So this seems to be some kind of length restriction. Do you know why things are written to `var/folders/nw/qczy4tjj333g09tcy197q3s40000gn/T` on your system instead of just `/tmp/`?

**Reply by mohamedawnallah** [reply]
I don't know actually I see these kind of issues for the docker containers engine because there are setting to disable saving/sharing data to `/var/folders'. That said running the test cases using `make check` wouldn't require docker engine to be installed. I've also spent considerable amount of time to know why this happen but actually failed. Also I have found similar discussion on Lightning Network related to the same issue but not answered yet:
https://github.com/lightningnetwork/lnd/discussions/7015

**Reply by mohamedawnallah** [reply]
@guggero That said I run the lnd node from the source and I didn't face issues. I'm wondering if fixing this issue is not necessary to contribute to lnd repo ?

**Reply by mohamedawnallah** [reply, marked as answer]
I've just solved the issue using the `sudo make check` command it seems a permission issue. Thanks so much @guggero for pointing me in the right direction 🙏. Now I'm gonna pick up good first issues

**Reply by guggero** [reply]
Okay, good to know! I'm also working on a fix for bitcoind 26 which generally doesn't seem to work with unix sockets currently. Should have the PR up today or so.

> I'm wondering if fixing this issue is not necessary to contribute to lnd repo?

No, as mentioned above, if you create a PR, the full CI pipeline will run on it, running all those checks for you. Though if you are a first-time contributor, someone needs to manually approve the run of the CI, so you can save a lot of time by running the checks on your local machine first.

**Reply by mohamedawnallah** [reply]
Thanks a lot for your follow-up!

**Reply by guggero** [reply]
Can you try https://github.com/lightningnetwork/lnd/pull/8273 if that allows you to run `make check` successfully on your machine now (without `sudo` which shouldn't be needed and hints at some issue with the way your machine or Golang installation is set up)?

**Reply by mohamedawnallah** [reply]
Hey @guggero sorry it seems I missed this notification. I'm gonna try it now and let you know!

**Reply by mohamedawnallah** [reply]
@guggero Based on the latest commit yet 435f0f9fa3d7318b2d59a755d5ac83c4039adea1 in #8273, the test case `github.com/lightningnetwork/lnd/routing/chainview` run successfully without `sudo`. 

That said the following test case `github.com/lightningnetwork/lnd/watchtower/wtclient` fails which didn't happen on the latest commit on the master branch 30348baedcd9e9b157c4cdbf7f0c322fff20cc68.
```
--- FAIL: TestClient (0.00s)
    --- FAIL: TestClient/can_remove_due_to_un-acked_updates_(no_client_restart) (1.46s)
        client_test.go:2315:
            	Error Trace:	/Users/mohamedawnallah/Desktop/lnd/watchtower/wtclient/client_test.go:2315
            	            				/Users/mohamedawnallah/Desktop/lnd/watchtower/wtclient/client_test.go:2566
            	Error:      	Received unexpected error:
            	            	address in use
            	Test:       	TestClient/can_remove_due_to_un-acked_updates_(no_client_restart)
    --- FAIL: TestClient/send_and_recv (15.40s)
        client_test.go:1365: Received 7/200
        client_test.go:1365: Received 14/200
        client_test.go:1365: Received 21/200
        client_test.go:1365: Received 30/200
        client_test.go:1365: Received 38/200
        client_test.go:1365: Received 46/200
        client_test.go:1365: Received 54/200
        client_test.go:1365: Received 64/200
        client_test.go:1365: Received 73/200
        client_test.go:1365: Received 82/200
        client_test.go:1365: Received 91/200
        client_test.go:1365: Received 99/200
        client_test.go:1365: Received 128/200
        client_test.go:1365: Received 154/200
        client_test.go:1365:
            	Error Trace:	/Users/mohamedawnallah/Desktop/lnd/watchtower/wtclient/client_test.go:1033
            	            				/Users/mohamedawnallah/Desktop/lnd/watchtower/wtclient/client_test.go:1365
            	            				/Users/mohamedawnallah/Desktop/lnd/watchtower/wtclient/client_test.go:2566
            	Error:      	Should be true
            	Test:       	TestClient/send_and_recv
            	Messages:   	breach hints not received, only got 178/200
    --- FAIL: TestClient/removing_a_tower_during_session_negotiation (15.60s)
        client_test.go:1815:
            	Error Trace:	/Users/mohamedawnallah/Desktop/lnd/watchtower/wtclient/client_test.go:1815
            	            				/Users/mohamedawnallah/Desktop/lnd/watchtower/wtclient/client_test.go:2566
            	Error:      	Received unexpected error:
            	            	predicate not satisfied after time out
            	Test:       	TestClient/removing_a_tower_during_session_negotiation
FAIL
FAIL	github.com/lightningnetwork/lnd/watchtower/wtclient	40.794s
```

## Additional Context
Bitcoin version used: `V23.0.0`.
Go Version: `go1.21.5 darwin/arm64`.
Go Binary: `/opt/homebrew/bin/go` installed using `brew install go` on Mac.

**Reply by guggero** [reply]
Okay, those are just flaky tests now. I'm sure if you re-run them they'll pass (and maybe others will fail)... It's hard to write super stable tests. If you find out what the issue is that makes them flakey, you'd be a hero!

**Reply by mohamedawnallah** [reply]
Perhaps I'll look into that soon! 😊
